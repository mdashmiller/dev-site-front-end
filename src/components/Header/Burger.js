import React from 'react'
import PropTypes from 'prop-types'

const Burger = ({ width }) => {
  // only render the burger button
  // on mobile screens
  if (width < 993) {
    return (
      <button data-test="burger-btn" className="burger">
        <i className="material-icons">menu</i>
      </button>
    )
  } else {
    return null
  }
}

Burger.propTypes = {
  width: PropTypes.number.isRequired
}

export default Burger
