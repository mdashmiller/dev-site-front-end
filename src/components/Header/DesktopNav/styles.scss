@import '../../../../Utils/master';

nav {
  ul a {
    color: $dark-pink;
    font-size: 1.1rem;
    font-weight: bold;

    &.btn-floating {
      background-color: $dark-pink;

      &:hover,
      &:focus {
        background-color: $darker-pink;
      }
    }

    &:hover {
      background-color: $darker-grey;
    }
  }
}
