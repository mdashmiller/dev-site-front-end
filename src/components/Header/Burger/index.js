import React from 'react'
import './styles.scss'

const Burger = () => {
  return (
    <button data-test="burger-btn" className="burger">
      <i data-test="menu-icon" className="material-icons">menu</i>
    </button>
  )
}

export default Burger
