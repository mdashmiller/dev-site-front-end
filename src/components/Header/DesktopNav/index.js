import React from 'react'
import './styles.scss'

const DesktopNav = () => {
  return (
    <ul data-test="links" className="right">
      <li><a data-test="about" href="#about">ABOUT</a></li>
      <li><a data-test="portfolio" href="#portfolio">PORTFOLIO</a></li>
      <li><a data-test="contact" href="#contact">CONTACT</a></li>
      <li>
        <a data-test="linkedin"
          href="linkedin.com"
          className="btn-floating btn-small tooltipped"
          data-tooltip="Linked In"
        >
          <i data-test="linkedin-icon" className="fab fa-linkedin"></i>
        </a>
      </li>
      <li>
        <a data-test="github"
          href="gitub.com"
          className="btn-floating btn-small tooltipped"
          data-tooltip="GitHub"
        >
          <i data-test="github-icon" className="fab fa-github"></i>
        </a>
      </li>
    </ul>
  )
}

export default DesktopNav
