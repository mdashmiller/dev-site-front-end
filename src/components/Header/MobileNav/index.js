import React, { Component } from 'react'
import './styles.scss'

class MobileNav extends Component {

  state = {
    mobileNavOpen: false
  }

  handleClick = () => {
    const { mobileNavOpen } = this.state
    this.setState({
      mobileNavOpen: !mobileNavOpen
    })
  }

  handleOutsideClick = (e) => {
    // close mobile nav menu if
    // user clicks away from it
    const { mobileNavOpen } = this.state
    if (mobileNavOpen) {
      if (!this.node.contains(e.target)) {
        this.setState({
          mobileNavOpen: false
        })
      }
    }
  }

  componentDidMount() {
    document.addEventListener('click', (e) => this.handleOutsideClick(e))
  }

  render() {
    const { mobileNavOpen } = this.state
    const toggle = mobileNavOpen ? 'visible' : null

    return (
      <div ref={node => this.node = node}>
        <button
          data-test="burger-btn"
          className="burger"
          onClick={(e) => this.handleClick(e)}
        >
          <i data-test="menu-icon" className="material-icons">menu</i>
        </button>
        <div
          data-test="menu"
          className={`mobile-menu ${toggle}`}
        >
          <ul>
            <li>
              <a
                data-test="about"
                href="#about"
                onClick={(e) => this.handleClick(e)}
              >
                About
              </a>
            </li>
            <li>
              <a
                data-test="portfolio"
                href="#portfolio"
                onClick={(e) => this.handleClick(e)}
              >
                Portfolio
              </a>
            </li>
            <li>
              <a
                data-test="contact"
                href="#contact"
                onClick={(e) => this.handleClick(e)}
              >
                Contact
              </a>
            </li>
          </ul>
        </div>
      </div>
    )
  }
}

export default MobileNav
