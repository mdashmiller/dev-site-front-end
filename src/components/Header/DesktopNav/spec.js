import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

import DesktopNav from './index'

import { findByTestAttr } from '../../../../Utils'

// render component for testing
const setUp = (props = {}) => {
  return shallow(<DesktopNav {...props} />)
}

describe('DesktopNav rendering', () => {

  let component
  beforeAll(() => {
    component = setUp()
  })

  it('should render 1 ul of links', () => {
    const wrapper = findByTestAttr(component, 'links')
    expect(wrapper.length).toBe(1)
  })

  it('should render 1 link to the about section', () => {
    const wrapper = findByTestAttr(component, 'about')
    expect(wrapper.length).toBe(1)
  })

  it('should render 1 link to the portfolio section', () => {
    const wrapper = findByTestAttr(component, 'portfolio')
    expect(wrapper.length).toBe(1)
  })

  it('should render 1 link to the contact section', () => {
    const wrapper = findByTestAttr(component, 'contact')
    expect(wrapper.length).toBe(1)
  })

  it('should render 1 link to linkedin', () => {
    const wrapper = findByTestAttr(component, 'linkedin')
    expect(wrapper.length).toBe(1)
  })

  it('should render 1 link to github', () => {
    const wrapper = findByTestAttr(component, 'github')
    expect(wrapper.length).toBe(1)
  })

  it('should render 1 linkedin icon', () => {
    const wrapper = findByTestAttr(component, 'linkedin-icon')
    expect(wrapper.length).toBe(1)
  })

  it('should render 1 github icon', () => {
    const wrapper = findByTestAttr(component, 'github-icon')
    expect(wrapper.length).toBe(1)
  })

})

describe('DesktopNav mounting and unmounting', () => {

  it('should render without error', () => {
    const div = document.createElement('div')
    ReactDOM.render(<DesktopNav />, div)
    ReactDOM.unmountComponentAtNode(div)
  })

})
