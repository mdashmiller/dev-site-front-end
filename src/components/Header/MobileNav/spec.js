import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

import MobileNav from './index'

import { findByTestAttr } from '../../../../Utils'

// render component for testing
const setUp = (props = {}) => {
  return shallow(<MobileNav {...props} />)
}

describe('MobileNav rendering', () => {

  let component
  beforeAll(() => {
    component = setUp()
  })

  it('should render 1 burger button', () => {
    const wrapper = findByTestAttr(component, 'burger-btn')
    expect(wrapper.length).toBe(1)
  })

  it('should render 1 menu icon', () => {
    const wrapper = findByTestAttr(component, 'menu-icon')
    expect(wrapper.length).toBe(1)
  })

  it('should render 1 menu div', () => {
    const wrapper = findByTestAttr(component, 'menu')
    expect(wrapper.length).toBe(1)
  })

  it('should render 1 link to the about section', () => {
    const wrapper = findByTestAttr(component, 'about')
    expect(wrapper.length).toBe(1)
  })

  it('should render 1 link to the portfolio section', () => {
    const wrapper = findByTestAttr(component, 'portfolio')
    expect(wrapper.length).toBe(1)
  })

  it('should render 1 link to the contact section', () => {
    const wrapper = findByTestAttr(component, 'contact')
    expect(wrapper.length).toBe(1)
  })

})

describe('conditional class "toggle"', () => {

  let component
  beforeEach(() => {
    component = setUp()
  })

  describe('state.mobileNavOpen is true', () => {

    it('should add className "visible" to menu div', () => {
      component.setState({ mobileNavOpen: true })
      const wrapper = findByTestAttr(component, 'menu')

      expect(wrapper.hasClass('visible')).toBe(true)
      expect(wrapper.hasClass('null')).toBe(false)
    })

  })

  describe('state.mobileNavOpen is false', () => {

    it('should add className "null" to menu div', () => {
      component.setState({ mobileNavOpen: false })
      const wrapper = findByTestAttr(component, 'menu')

      expect(wrapper.hasClass('visible')).toBe(false)
      expect(wrapper.hasClass('null')).toBe(true)
    })

  })

})

describe('handleClick', () => {
  
  let component, instance
  beforeEach(() => {
    component = setUp()
    instance = component.instance()
    jest.spyOn(instance, 'handleClick')
  })

  describe('spying on handleClick', () => {

    it('should be called when burger button is clicked', () => {
      const burger = findByTestAttr(component, 'burger-btn')
      burger.simulate('click')

      expect(instance.handleClick).toHaveBeenCalledTimes(1)
    })

    it('should be called when a link is clicked', () => {
      const about = findByTestAttr(component, 'about')
      const portfolio = findByTestAttr(component, 'portfolio')
      const contact = findByTestAttr(component, 'contact')

      about.simulate('click')
      portfolio.simulate('click')
      contact.simulate('click')

      expect(instance.handleClick).toHaveBeenCalledTimes(3)
    })

  })

  describe('directly invoking handleClick', () => {

    it('should toggle state.mobileNavOpen', () => {
      instance.handleClick()
      expect(component.state('mobileNavOpen')).toBe(true)

      instance.handleClick()
      expect(component.state('mobileNavOpen')).toBe(false)
    })

    it('should add the class "visible" to the mobile-menu ul', () => {
      let menu = findByTestAttr(component, 'menu')
      expect(menu.hasClass('visible')).toBe(false)

      instance.handleClick()
      
      menu = findByTestAttr(component, 'menu')
      expect(menu.hasClass('visible')).toBe(true)
    })

  })

})

describe('handleOutsideClick()', () => {

  let component, instance
  beforeEach(() => {
    component = setUp()
    instance = component.instance()
  })

  // describe('spying on handleOutsideClick()', () => {

  //   it('should be called when any part of document is clicked', () => {
  //     jest.spyOn(instance, 'handleOutsideClick')

  //   })

  // })

  describe('directly invoking handleOutsideClick()', () => {

    describe('mobile-nav is not visible', () => {

      it('should not change state.mobileNavOpen', () => {
        component.setState({ mobileNavOpen: false })
        instance.handleOutsideClick()
        expect(component.state('mobileNavOpen')).toBe(false)
      })

    })

    // describe('mobile-nav is visible', () => {

    //   describe('user clicks the mobile-nav', () => {

    //     it('should not change state.mobileNavOpen', () => {

    //     })

    //   })

    //   describe('user clicks anything other than mobile-nav', () => {

    //     it('should set state.mobileNavOpen to false', () => {
          
    //     })

    //   })


    // })

  })

})

// describe('componentDidMount()', () => {

//   it('should attach a "click" event listener to the document', () => {

//   })

// })

describe('MobileNav mounting and unmounting', () => {

  it('should render without error', () => {
    const div = document.createElement('div')
    ReactDOM.render(<MobileNav />, div)
    ReactDOM.unmountComponentAtNode(div)
  })

})
