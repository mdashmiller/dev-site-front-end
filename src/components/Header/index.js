import React, { Component } from 'react'
import './styles.scss'
import logo from '../../assets/images/logo.jpg'
import DesktopNav from './DesktopNav/index'
import MobileNav from './MobileNav/index'

class Header extends Component {

  state = {
    windowWidth: window.innerWidth
  }

  setWindowWidth = () => {
    this.setState({
      windowWidth: window.innerWidth
    })
  }

  componentDidMount() {
    // track changes in browser width
    window.addEventListener('resize', () => this.setWindowWidth())
  }

  render() {
    const { windowWidth } = this.state
    
    return (
      <header data-test="header" id="top" className="scrollspy">
        <nav data-test="nav">
          <div data-test="nav-wrapper" className="nav-wrapper container">
            <a href="#top" data-test="brand-logo" className="brand-logo">
              <img src={logo} data-test="logo-img" alt="Matt Miller Web Developer" />
            </a>
            { windowWidth > 992 && <DesktopNav data-test="desktop-nav" /> }
          </div>
        </nav>
        { windowWidth < 993 && <MobileNav data-test="mobile-nav" /> }
      </header>
    )
  }
}

export default Header
