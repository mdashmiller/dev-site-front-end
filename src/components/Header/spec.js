import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

import Header from './index'
import DesktopNav from './DesktopNav/index'
import MobileNav from './MobileNav/index'

import { findByTestAttr } from '../../../Utils'

// render component for testing
const setUp = (props={}) => {
  return shallow(<Header {...props} />)
}

describe('Header rendering', () => {

  let component
  beforeEach(() => {
    component = setUp()
  })

  it('should render 1 header', () => {
    const wrapper = findByTestAttr(component, 'header')
    expect(wrapper.length).toBe(1)
  })

  it('should render 1 nav', () => {
    const wrapper = findByTestAttr(component, 'nav')
    expect(wrapper.length).toBe(1)
  })

  it('should render 1 nav-wrapper div', () => {
    const wrapper = findByTestAttr(component, 'nav-wrapper')
    expect(wrapper.length).toBe(1)
  })

  it('should render 1 brand logo', () => {
    const wrapper = findByTestAttr(component, 'brand-logo')
    expect(wrapper.length).toBe(1)
  })

  it('should render 1 logo img', () => {
    const wrapper = findByTestAttr(component, 'logo-img')
    expect(wrapper.length).toBe(1)
  })

  describe('mobile specific rendering', () => {

    // change viewport to mobile size
    beforeAll(() => {
      global.innerWidth = 992
      global.dispatchEvent(new Event('resize'))
    })

    it('should render 1 MobileNav component', () => {
      const wrapper = findByTestAttr(component, 'mobile-nav')
      expect(wrapper.length).toBe(1)
    })

    it('should not render the DesktopNav component', () => {
      const wrapper = findByTestAttr(component, 'desktop-nav')
      expect(wrapper.length).toBe(0)
    })

  })

  describe('desktop specific rendering', () => {

    // change viewport to desktop size
    beforeAll(() => {
      global.innerWidth = 993
      global.dispatchEvent(new Event('resize'))
    })

    it('should render 1 DesktopNav component', () => {
      const wrapper = findByTestAttr(component, 'desktop-nav')
      expect(wrapper.length).toBe(1)
    })

    it('should not render the MobileNav component', () => {
      const wrapper = findByTestAttr(component, 'mobile-nav')
      expect(wrapper.length).toBe(0)
    })

  })

})

describe('setWindowWidth()', () => {

  let component
  let instance
  beforeEach(() => {
    component = setUp()
    instance = component.instance()
  })

  describe('spying on setWindowWidth()', () => {
    
    it('should be called when the browser window is resized', () => {
      jest.spyOn(instance, 'setWindowWidth')

      // resize browser window
      global.innerWidth = 1000
      global.dispatchEvent(new Event('resize'))

      expect(instance.setWindowWidth).toHaveBeenCalledTimes(1)
    })

  })

  describe('directly invoking setWindowWidth()', () => {

    it('should set state.windowWidth with value of window.innerWidth', () => {
      global.innerWidth = 800

      instance.setWindowWidth()
      
      expect(component.state('windowWidth')).toBe(800)
    })

  })

})

// describe('componentDidMount()', () => {

//   describe('directly invoking componentDidMount()', () => {

//     it('should attach the "resize" event listener to the window', () => {

//     })

//   })

// })

describe('Header mounting and unmounting', () => {

  it('should render without error', () => {
    const div = document.createElement('div')
    ReactDOM.render(<Header />, div)
    ReactDOM.unmountComponentAtNode(div)
  })

})

// describe('Header snapshot', () => {

//   it('should have a valid snapshot', () => {
//     const component = renderer.create(<Header />)
//     let tree = component.toJSON()
//     expect(tree).toMatchSnapshot()
//   })

// })
