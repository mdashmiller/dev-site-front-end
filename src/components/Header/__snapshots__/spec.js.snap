// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header snapshot should have a valid snapshot 1`] = `
<header
  className="scrollspy"
  data-test="header"
>
  <nav
    data-test="nav"
  >
    <div
      className="nav-wrapper container"
      data-test="nav-wrapper"
    >
      <a
        className="brand-logo"
        data-test="brand-logo"
        href="#top"
      >
        <img
          alt="Matt Miller Web Developer"
          data-test="logo-img"
          src="logo.jpg"
        />
      </a>
      <ul
        className="right"
        data-test="links"
      >
        <li>
          <a
            data-test="about"
            href="#about"
          >
            ABOUT
          </a>
        </li>
        <li>
          <a
            data-test="portfolio"
            href="#portfolio"
          >
            PORTFOLIO
          </a>
        </li>
        <li>
          <a
            data-test="contact"
            href="#contact"
          >
            CONTACT
          </a>
        </li>
        <li>
          <a
            className="btn-floating btn-small tooltipped"
            data-test="linkedin"
            data-tooltip="Linked In"
            href="linkedin.com"
          >
            <i
              className="fab fa-linkedin"
              data-test="linkedin-icon"
            />
          </a>
        </li>
        <li>
          <a
            className="btn-floating btn-small tooltipped"
            data-test="github"
            data-tooltip="GitHub"
            href="gitub.com"
          >
            <i
              className="fab fa-github"
              data-test="github-icon"
            />
          </a>
        </li>
      </ul>
    </div>
  </nav>
</header>
`;
